frontend/schema/schema.go

```diff
@@ -0,0 +1,298 @@
+/*
+Copyright © 2022 ConsenSys Software Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schema
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/consensys/gnark/internal/backend/compiled"
+)
+
+// Schema represents the structure of a gnark circuit (/ witness)
+type Schema []Field
+
+// LeafHandler is the handler function that will be called when Visit reaches leafs of the struct
+type LeafHandler func(visibility compiled.Visibility, name string, tValue reflect.Value) error
+
+// Parse filters recursively input data struct and keeps only the fields containing slices, arrays of elements of
+// type frontend.Variable and return the corresponding Schema. Slices are converted to arrays.
+//
+// If handler is specified, handler will be called on each encountered leaf (of type tLeaf)
+func Parse(circuit interface{}, tLeaf reflect.Type, handler ...LeafHandler) (Schema, error) {
```

Comment:
no handler is not set when you just want the schema (for now, only in schema_test.go) . We can put it as arg and check if it's nil to not call it, works well too. 

For the tLeaf, indeed, it's to avoid import cycles, (same as for why circuit is interface{} and not frontend.Circuit)